service/validation.go

```diff
@@ -1,97 +1,117 @@
 package service
 
 import (
+	"errors"
 	"fmt"
+
+	"github.com/mesg-foundation/core/utils/vtree"
 )
 
-// ParameterWarning contains a specific warning related to a parameter.
-type ParameterWarning struct {
-	Key       string
-	Warning   string
-	Parameter *Parameter
+type parameterValidator struct {
+	warnings []*ParameterWarning
 }
 
-func (p *ParameterWarning) String() string {
-	return fmt.Sprintf("Value of %q is %s", p.Key, p.Warning)
+func newParameterValidator() *parameterValidator {
+	return &parameterValidator{warnings: make([]*ParameterWarning, 0)}
 }
 
-// ValidateParametersSchema validates data to see if it matches with parameters schema.
+// Validate validates data to check if it matches with parameter's schema.
 // TODO(ilgooz) add this as a method to Service type when create custom types for Event, Task etc.
 // TODO(ilgooz) remove pointer from *Parameter.
-func validateParametersSchema(parameters []*Parameter, data map[string]interface{}) []*ParameterWarning {
-	warnings := make([]*ParameterWarning, 0, len(parameters))
-
-	for _, param := range parameters {
-		warning := newParameterValidator(param).Validate(data[param.Key])
-		if warning != nil {
-			warnings = append(warnings, warning)
-		}
+func (p *parameterValidator) Validate(params []*Parameter, data interface{}) ([]*ParameterWarning, error) {
+	v := vtree.Analyze(data)
```

Comment:
Doing validation by using reflection is not efficient.
Reflection should not be use in high demanded function.
Here, for every event, result, task that the code receive, it will do reflection.

The goal is to do the reflection once to populate a structure that can do the validation efficiently.

Reply:
The reason why we use reflection is to support struct types and not just maps. We can always change the implementation of _vtree_ pkg to do type assertions instead of using reflection. But that way we're no longer be able to support structs.

_About why we need to support structs:_

In system services it's easier to pass existent struct types as input data. In the [example](https://github.com/mesg-foundation/core/blob/dev/systemservices/workflow/workflow.go#L20) below we don't have this flexibility.

We do a small hack here by using json marshal & unmarshal to convert struct type to a map in order to satisfy old implementation of the parameter validation func because it can only accept map types. Since json pkg heaviliy depends on reflection pkg to analyze given data type and set values it got from json string to given type this operation is very heavy. Wherareas _vtree_ pkg only analyzes the data type, it doesn't do further reflection like json.Unmarshal does to set values and we only call Analyze method once. So this can be considered as a lightweight operation and a small giveaway in favor of supporting structs.
```go

func CreateInputs(definition workflowparser.WorkflowDefinition, name string) (map[string]interface{}, error) {
	// TODO: this hack is not something that we should do but
	// it's needed because *parameterValidator is not able to identify
	// structs for now.
	definitionData, err := json.Marshal(definition)
	if err != nil {
		return nil, err
	}
	var data interface{}
	if err := json.Unmarshal(definitionData, &data); err != nil {
		return nil, err
	}
	return map[string]interface{}{
		"definition": data,
		"name":       name,
	}, nil
}
``` 

So with the power of _vtree_ pkg this code becomes to (should x3-x4 faster because less reflection made):

```go
func CreateInputs(definition workflowparser.WorkflowDefinition, name string) (map[string]interface{}, error) {
    return map[string]interface{}{
        "definition": definition,
        "name":       name,
    }, nil
}
```

We can also change the implementation of _vtree_ to drop struct support and reflection and only do type assertion but in that case we need to manually convert struct types to maps and this will cause to long, hart to read code and specially if the struct consist of more nested structs. It can quickly become a nightmare and it's hard to maintain.

Reply:
One other option is to change the implementation of [ExecuteTask](https://github.com/mesg-foundation/core/blob/2cd3fa0/api/execute.go#L31) to accept json encoded byte data instead of a map type. This way we still going to be able to use all kind of Go types including structs in the system services but we need to encode them to json before passing to the _ExecuteTask_.

Then _ExecuteTask_ can unmarshal it to a map and we can to type assertion on top of that inside the _vtree_ pkg.

Pros:
* No need for doing reflection ourselves within the _vtree_.

Cons:
* We limit us using json as the encoding type by requiring to pass byte representation of it to the _ExecuteTask_ func. Maybe not a big cons.

* For system services we'll need to marshal to json and unmarshal it again inside the _ExecuteTask_ for no real reason which lowers the performance x2 time where we could just use the data type as it is and stick with _vtree_ where it'll do a small amount of reflection.

Reply:
here is a sample [PR](https://github.com/mesg-foundation/core/pull/657) when assertion is adopted within _vtree_.